### PR TITLE
[fix][proxy] Refresh auth data if ProxyLookupRequests

### PR DESCRIPTION
**Pull request by michaeljmarshall**
Fixes: https://github.com/apache/pulsar/issues/10816
PIP: #19771
Supersedes: https://github.com/apache/pulsar/pull/19026
Depends on: https://github.com/apache/pulsar/pull/20062

### Motivation

The Pulsar Proxy does not properly handle authentication data refresh when in state `ProxyLookupRequests`. The consequence is described in https://github.com/apache/pulsar/issues/10816. Essentially, the problem is that the proxy caches stale authentication data and sends it to the broker leading to connection failures.

https://github.com/apache/pulsar/pull/17831 attempted to fix the underlying problem, but it missed an important edge cases. Specifically, it missed the case that the `ConnectionPool` will have multiple connections when a lookup gets redirected. As such, the following problem exists (and is fixed by this PR):

1. Client opens connection to perform lookups.
2. Proxy connects to broker 1 to get the topic ownership info.
3. Time passes.
4. Client does an additional lookup, and this topic is on a newly created broker 2. In this case, the proxy opens a new connection with the stale client auth data.
5. Broker 2 rejects the connection because it fails with expired authentication.

### Modifications

* Remove some of the implementation from https://github.com/apache/pulsar/pull/17831. This new implementation still allows a broker to challenge the client through the proxy, but notably, it limits the number of challenges sent to the client. Further, the proxy does not challenge the client when the auth data is not expired.
* Introduce authentication refresh in the proxy so that the proxy challenges the client any time the auth data is expired.
* Update the `ProxyClientCnx` to get the `clientAuthData` from the `ProxyConnection` to ensure that it gets new authentication data.
* Add clock skew to the `AuthenticationProviderToken`. This is necessary to make some of the testing not flaky and it will also be necessary for users to configure in their clusters.

### Verifying this change

The `ProxyRefreshAuthTest` covers the existing behavior and I expanded it to cover the edge case described above.

Additionally, testing this part of the code will be much easier to test once we implement #19624.

### Documentation

- [x] `doc-not-needed`

### Matching PR in forked repository

PR in forked repository: the relevant tests pass locally, so I am going to skip the forked tests.

### CODE REVIEW

**Review by lhotari — APPROVED** [review]
LGTM

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java [hunk 1]
```diff
@@ -471,6 +497,47 @@ protected void authChallengeSuccessCallback(AuthData authChallenge) {
...
+
+        if (System.nanoTime() - authChallengeSentTime
+                > TimeUnit.SECONDS.toNanos(service.getConfiguration().getAuthenticationRefreshCheckSeconds())) {
+            LOG.warn("[{}] Closing connection after timeout on refreshing auth credentials", remoteAddress);
+            ctx.close();
+            return;
+        }
+
+        if (LOG.isDebugEnabled()) {
+            LOG.info("[{}] Refreshing authentication credentials", remoteAddress);
```
Comment (nicoloboschi) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java:524]: LOG.debug ?

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java [hunk 2]
```diff
@@ -376,6 +395,14 @@ private synchronized void completeConnect() throws PulsarClientException {
 
             state = State.ProxyLookupRequests;
             lookupProxyHandler = service.newLookupProxyHandler(this);
+            if (service.getConfiguration().isAuthenticationEnabled()
+                    && service.getConfiguration().getAuthenticationRefreshCheckSeconds() > 0) {
+                authRefreshTask = ctx.executor().scheduleAtFixedRate(
+                        Runnables.catchingAndLoggingThrowables(() -> refreshAuthenticationCredentials(false)),
```
Comment (nodece) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java:401]: It will never refresh the authentication credential.

`(!authState.isExpired() || !force)` always `true`.

Comment (michaeljmarshall) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java:401]: Great catch, that is a bad mistake. I'll fix it now.

**Review by nodece — APPROVED** [review]
LGTM

**Comment by michaeljmarshall** [comment]
@cbornet - I think this is a necessary bug fix for 3.0. Without it, OIDC will not work correctly through the pulsar proxy. This bug was not introduced recently, but it didn't matter until now because the OIDC work exposes the bug. I've been fighting with CI all day due to flaky tests (and some legitimate failures, too). If we're able to hold off on creating `branch-3.0` just a little longer, I'd really appreciate it. Thanks!
